commit a59fd063e30e2d613c5751aeac468ea4211c888a (refs/changes/89/1489/2)
Author: Joshua M. Clulow <jmc@joyent.com>
Date:   2017-02-15T21:37:43+00:00 (2 years, 8 months ago)
    
    MANTA-3157 muskie picker ought to start even when there is no space

diff --git a/lib/picker.js b/lib/picker.js
index b2e797c..70741fa 100644
--- a/lib/picker.js
+++ b/lib/picker.js
@@ -241,8 +241,16 @@ function poll() {
                 count++;
             });
 
-            // Don't replace if we got an empty "DB"
-            if (count > 0) {
+            /*
+             * In general, we don't want to replace a non-empty set of sharks
+             * with an empty one.  We make an exception at startup: provided
+             * the query to Moray did not result in any errors, we accept one
+             * empty result set.  This enables us to service requests that will
+             * not ask the Picker for a storage node, while remaining confident
+             * that once storage nodes become available, we'll know about them.
+             */
+            if (count > 0 || self.firstTopology) {
+                self.firstTopology = false;
                 self.datacenters = dcs;
                 self.db = obj;
                 self.emit('topology', self.db);
@@ -354,6 +362,8 @@ function Picker(opts) {
     this.ignoreSize = opts.ignoreSize === undefined ? false : opts.ignoreSize;
 
     this.client.once('connect', poll.bind(this));
+
+    this.firstTopology = true;
     this.once('topology', this.emit.bind(this, 'connect'));
 }
 util.inherits(Picker, EventEmitter);
